c7n/resources/iam.py

```diff
@@ -465,6 +465,59 @@ def process(self, resources, event=None):
         return []
 
 
+@Role.action_registry.register('set-policy')
+class SetPolicy(BaseAction):
+    """Set a specific IAM policy as attached or detached on a role.
+
+    You will identify the policy by its arn.
+
+    Returns a list of roles modified by the action.
+
+    For example, if you want to automatically attach a policy to all roles which don't have it...
+
+    :example:
+
+      .. code-block:: yaml
+
+        - name: iam-attach-role-policy
+          resource: iam-role
+          filters:
+            - type: no-specific-managed-policy
+              value: my-iam-policy
+          actions:
+            - type: set-policy
+              state: attached
+              arn: arn:aws:iam::123456789012:policy/my-iam-policy
+
+    """
+    schema = type_schema('set-policy', state={'type': 'string'}, arn={'type': 'string'})
```

Comment:
for simpler validation then a validate method, you can just do 

```
{'state': 'enum': ['attached', 'detached']}
```
and also required=['state', 'arn'] as an argument to the type_schema function, and you could remove the validate method.

and then remove the validate method.

Reply:
also just thinking out loud, but we don't technically need an arn here, we could just work off policy name and generate the arn if not passed in arn format. we could then just rename 'arn' to 'policy'

Reply:
the arn turns out to be ambigious to generate due to pathing and some use of global sans arn.

Reply:
the arn turns out to be ambigious to generate due to pathing and some use of global sans arn.